src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extenstions.cs

```diff
@@ -0,0 +1,186 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+
+namespace Microsoft.AspNet.Server.Kestrel.Infrastructure
+{
+    public static class MemoryPoolIterator2Extenstions
+    {
+        private const int _maxStackAllocBytes = 16384;
+
+        private static Encoding _utf8 = Encoding.UTF8;
+        private static Encoding _ascii = Encoding.ASCII;
+
+        private static unsafe string GetAsciiStringStack(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
+        {
+            // avoid declaring other local vars, or doing work with stackalloc
+            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
+            char* output = stackalloc char[length];
+
+            return GetAsciiStringImplementation(output, start, end, inputOffset, length);
+        }
+
+        private unsafe static string GetAsciiStringHeap(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
+        {
+            var buffer = new char[length];
+            fixed (char* output = buffer)
+            {
+                return GetAsciiStringImplementation(output, start, end, inputOffset, length);
+            }
+        }
+
+        private static unsafe string GetAsciiStringImplementation(char* output, MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
+        {
+            var outputOffset = 0;
+            var block = start;
+            var remaining = length;
+
+            var endBlock = end.Block;
+            var endIndex = end.Index;
+
+            while (true)
+            {
+                int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+
+                if (following > 0)
+                {
+                    var input = block.Array;
+                    for (var i = 0; i < following; i++)
+                    {
+                        output[i + outputOffset] = (char)input[i + inputOffset];
```

Comment:
We don't have a good mechanism for treating it as `opaque data` because we only expose strings and you have to convert to a string somehow. If there was a cheap way to detect bad data I'd prefer to do a verbose log before we mangled it.